File: python/examples/native/search/search_factory.py
#!/usr/bin/env python3
"""
Search Tool - Factory Pattern with Interactive Approval

Example: Search for information in files using Matimo's search tool
Demonstrates searching files for patterns and content with interactive approval
"""

import asyncio
import os
import sys
from pathlib import Path

from dotenv import load_dotenv

from matimo import Matimo, get_global_approval_handler


def create_approval_callback():
    """Create an interactive approval callback for file operations"""

    async def approval_handler(request):
        """
        Handle approval requests interactively
        
        Args:
            request: ApprovalRequest with toolName, description, params
            
        Returns:
            bool: True if approved, False otherwise
        """
        is_interactive = sys.stdin.isatty()

        print('\n' + '=' * 70)
        print('🔒 APPROVAL REQUIRED FOR SEARCH OPERATION')
        print('=' * 70)
        print(f'\n📋 Tool: {request["toolName"]}')
        print(f'📝 Description: {request.get("description", "(no description provided)")}')
        print('\n🔍 Search Operation:')
        print(f'   Query: {request["params"].get("query", "N/A")}')
        print(f'   Directory: {request["params"].get("directory", "N/A")}')
        print(f'   Pattern: {request["params"].get("filePattern", "N/A")}')

        if not is_interactive:
            print('\n❌ REJECTED - Non-interactive environment (no terminal)')
            print('\n💡 To enable auto-approval in CI/scripts:')
            print('   export MATIMO_AUTO_APPROVE=true')
            print('\n💡 Or approve specific patterns:')
            print('   export MATIMO_APPROVED_PATTERNS="search"')
            print('\n' + '=' * 70 + '\n')
            return False

        # Interactive mode: prompt user
        print('\n❓ User Action Required')
        user_input = input('   Type "yes" to approve or "no" to reject: ').strip().lower()
        approved = user_input in ('yes', 'y')

        if approved:
            print('   ✅ Operation APPROVED by user')
        else:
            print('   ❌ Operation REJECTED by user')
        print('=' * 70 + '\n')

        return approved

    return approval_handler


async def main():
    """
    Main entry point for search factory example
    
    Demonstrates:
    1️⃣ Searching for "import" patterns in examples
    2️⃣ Searching for function definitions in core packages
    3️⃣ Searching for console patterns with regex
    """
    # Load environment variables
    load_dotenv()

    # Initialize Matimo with auto-discovery to find all tools
    matimo = await Matimo.init(auto_discover=True)

    # Configure centralized approval handler
    approval_handler = get_global_approval_handler()
    approval_handler.set_approval_callback(create_approval_callback())

    print('=== Search Tool - Factory Pattern (Interactive Approval) ===\n')

    # Get the workspace root (parent of examples directory)
    current_file = Path(__file__)
    workspace_root = current_file.parent.parent.parent.parent.parent

    try:
        # 1️⃣ Example 1: Search for pattern in TypeScript files
        print('1️⃣ Searching for "import" in examples\n')
        examples_dir = workspace_root / 'typescript' / 'examples' / 'tools'
        
        try:
            result1 = await matimo.execute(
                'search',
                {
                    'query': 'import',
                    'directory': str(examples_dir),
                    'filePattern': '*.ts',
                    'maxResults': 5,
                }
            )

            if result1.get('success'):
                print(f'Total matches: {result1.get("totalMatches")}')
                print(f'Matches found: {len(result1.get("matches", []))}')
                
                # Show first 3 matches
                for match in result1.get('matches', [])[:3]:
                    print(f'  - {match.get("filePath")}:{match.get("lineNumber")}: {match.get("lineContent")}')
            else:
                print(f'Search denied: {result1.get("error")}')
        except Exception as e:
            print(f'Error: {e}')
        
        print('---\n')

        # 2️⃣ Example 2: Search for function definitions
        print('2️⃣ Searching for function definitions\n')
        core_src_dir = workspace_root / 'typescript' / 'packages' / 'core' / 'src'
        
        try:
            result2 = await matimo.execute(
                'search',
                {
                    'query': 'export function',
                    'directory': str(core_src_dir),
                    'filePattern': '*.ts',
                    'maxResults': 10,
                }
            )

            if result2.get('success'):
                print(f'Total matches: {result2.get("totalMatches")}')
                print(f'Matches found: {len(result2.get("matches", []))}')
            else:
                print(f'Search denied: {result2.get("error")}')
        except Exception as e:
            print(f'Error: {e}')
        
        print('---\n')

        # 3️⃣ Example 3: Search in specific directory with regex
        print('3️⃣ Searching for "console" patterns with regex\n')
        
        try:
            result3 = await matimo.execute(
                'search',
                {
                    'query': r'console\.log',
                    'directory': str(core_src_dir),
                    'filePattern': '*.ts',
                    'isRegex': True,
                    'maxResults': 5,
                }
            )

            if result3.get('success'):
                print(f'Total matches: {result3.get("totalMatches")}')
                print(f'Matches found: {len(result3.get("matches", []))}')
            else:
                print(f'Search denied: {result3.get("error")}')
        except Exception as e:
            print(f'Error: {e}')
        
        print('---\n')

    except Exception as error:
        print(f'Error searching files: {error}')
        if hasattr(error, 'message'):
            print(f'Error message: {error.message}')


if __name__ == '__main__':
    asyncio.run(main())

File: typescript/examples/tools/search/search-factory.ts
import { MatimoInstance, getGlobalApprovalHandler, type ApprovalRequest } from '@matimo/core';
import path from 'path';
import { fileURLToPath } from 'url';
import * as readline from 'readline';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

/**
 * Create an interactive approval callback for file operations
 */
function createApprovalCallback() {
  return async (request: ApprovalRequest): Promise<boolean> => {
    const isInteractive = process.stdin.isTTY;

    console.info('\n' + '='.repeat(70));
    console.info('🔒 APPROVAL REQUIRED FOR FILE OPERATION');
    console.info('='.repeat(70));
    console.info(`\n📋 Tool: ${request.toolName}`);
    console.info(`📝 Description: ${request.description || '(no description provided)'}`);
    console.info(`\n📄 File Operation:`);
    console.info(`   Path: ${request.params.filePath}`);
    if (request.params.startLine) {
      console.info(`   Start Line: ${request.params.startLine}`);
    }
    if (request.params.endLine) {
      console.info(`   End Line: ${request.params.endLine}`);
    }

    if (!isInteractive) {
      console.info('\n❌ REJECTED - Non-interactive environment (no terminal)');
      console.info('\n💡 To enable auto-approval in CI/scripts:');
      console.info('   export MATIMO_AUTO_APPROVE=true');
      console.info('\n💡 Or approve specific patterns:');
      console.info('   export MATIMO_APPROVED_PATTERNS="search"');
      console.info('\n' + '='.repeat(70) + '\n');
      return false;
    }

    // Interactive mode: prompt user
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });

    return new Promise((resolve) => {
      console.info('\n❓ User Action Required');
      const question = '   Type "yes" to approve or "no" to reject: ';

      rl.question(question, (answer) => {
        const approved = answer.toLowerCase() === 'yes' || answer.toLowerCase() === 'y';

        if (approved) {
          console.info('   ✅ Operation APPROVED by user');
        } else {
          console.info('   ❌ Operation REJECTED by user');
        }
        console.info('='.repeat(70) + '\n');

        rl.close();
        resolve(approved);
      });
    });
  };
}

/**
 * Example: Search tool using factory pattern
 * Demonstrates searching files for patterns and content with interactive approval
 */
async function searchExample() {
  // Initialize Matimo with autoDiscover to find all tools (core + providers)
  const matimo = await MatimoInstance.init({ autoDiscover: true });

  // Configure centralized approval handler
  const approvalHandler = getGlobalApprovalHandler();
  approvalHandler.setApprovalCallback(createApprovalCallback());

  console.info('=== Search Tool - Factory Pattern (Interactive Approval) ===\n');

  // Get the workspace root (parent of examples directory)
  // File is in examples/tools/search/, so go up 3 levels
  const workspaceRoot = path.resolve(__dirname, '../../..');

  try {
    // Example 1: Search for pattern in TypeScript files
    console.info('1. Searching for "import" in examples\n');
    const result1 = await matimo.execute('search', {
      query: 'import',
      directory: path.join(workspaceRoot, 'examples/tools'),
      filePattern: '*.ts',
      maxResults: 5,
    });

    if ((result1 as any).success) {
      console.info('Total matches:', (result1 as any).totalMatches);
      console.info('Matches found:', (result1 as any).matches?.length);
      if ((result1 as any).matches) {
        (result1 as any).matches.slice(0, 3).forEach((match: any) => {
          console.info(`  - ${match.filePath}:${match.lineNumber}: ${match.lineContent}`);
        });
      }
    } else {
      console.info('Search denied:', (result1 as any).error);
    }
    console.info('---\n');

    // Example 2: Search for function definitions
    console.info('2. Searching for function definitions\n');
    const result2 = await matimo.execute('search', {
      query: 'export function',
      directory: path.join(workspaceRoot, 'packages/core/src'),
      filePattern: '*.ts',
      maxResults: 10,
    });

    if ((result2 as any).success) {
      console.info('Total matches:', (result2 as any).totalMatches);
      console.info('Matches found:', (result2 as any).matches?.length);
    } else {
      console.info('Search denied:', (result2 as any).error);
    }
    console.info('---\n');

    // Example 3: Search in specific directory with regex
    console.info('3. Searching for "console.info" patterns\n');
    const result3 = await matimo.execute('search', {
      query: 'console\\.info',
      directory: path.join(workspaceRoot, 'packages/core/src'),
      filePattern: '*.ts',
      isRegex: true,
      maxResults: 5,
    });

    if ((result3 as any).success) {
      console.info('Total matches:', (result3 as any).totalMatches);
      console.info('Matches found:', (result3 as any).matches?.length);
    } else {
      console.info('Search denied:', (result3 as any).error);
    }
    console.info('---\n');
  } catch (error: any) {
    console.error('Error searching files:', error.message);
  }
}

searchExample();
